### PR TITLE
Fix Auth#isStaff

### PR DESCRIPTION
**Pull request by iscke**

### CODE REVIEW

**Review by AnnikaCodes — APPROVED** [review]

File: server/user-groups.ts [hunk 1]
```diff
@@ -56,7 +56,9 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
 		return super.get(user) || Auth.defaultSymbol();
 	}
 	isStaff(userid: ID) {
-		return this.has(userid) && this.get(userid) !== '+';
+		if (!this.has(userid)) return false;
+		const auth = this.get(userid);
+		return auth !== '+' && auth !== Auth.defaultSymbol();
```
Comment (Zarel) [inline server/user-groups.ts:61]: Oh, we need to hardcode Prize Winner now.

```suggestion
		return auth !== '+' && auth !== '^' && auth !== Auth.defaultSymbol();
```

Maybe the individual config entries should mark `isStaff`?

**Comment by whalemer** [comment]
how's this?

**Comment by Zarel** [comment]
I wonder if it should be an actual permission. We also use `lock` for a lot of permissions that aren't actually related to locks...

**Comment by Zarel** [comment]
Alternatively, just do the `^` hardcode, for now?

**Comment by whalemer** [comment]
having it as a proper permission feels like a misname. what does it give you that you can't already do through Auth#isStaff and setting it manually in config?

(i'm pretty sure this needs a restart to take effect, so i'm not too worried about pushing it through particularly quickly)

File: server/user-groups.ts [hunk 2]
```diff
@@ -89,9 +90,9 @@ export abstract class Auth extends Map<ID, GroupSymbol | ''> {
 			name: symbol,
 		});
 	}
-	getEffectiveSymbol(user: User): EffectiveGroupSymbol {
+	getEffectiveSymbol(user: ID | User): EffectiveGroupSymbol {
 		const group = this.get(user);
-		if (this.has(user.id) && group === Auth.defaultSymbol()) {
+		if (this.has(typeof user !== 'string' ? (user as User).id : user) && group === Auth.defaultSymbol()) {
```
Comment (AnnikaCodes) [inline server/user-groups.ts:95]:
```suggestion
		if (this.has(toID(user)) && group === Auth.defaultSymbol()) {
```

**Comment by AnnikaCodes** [comment]
I think having it in the same place as `roomonly` makes sense - it's metadata about the group itself, rather than a permission in and of itself.

**Comment by Zarel** [comment]
I think the first question is "what is it used for"? Like `lock`, I think there's an argument that a lot of those things should be individual permissions.

**Comment by Zarel** [comment]
If we have time, let's create a list. My recent refactor migrated IP-viewing from `ban` to `ip`, which was a good start.

Things left:

- Viewing private mod actions
- Autojoining Staff?

Can one of you make a list?

**Comment by Zarel** [comment]
- Viewing private mod actions
- Autojoining Staff
- Automatically grant "trusted"
- use STAFF_THROTTLE rather than regular throttle
- A quotes permission hardcode which seems unnecessary?
- Allow certain monitor bypasses
- Allow use of the helpticket system

That seems to be it for `user.isStaff`.

**Comment by Zarel** [comment]
`auth.isStaff` is used for:

- `STAFF_THROTTLE` again
- trusted user checking
- chatfilter bypassing
- a Mafia chatfilter bypass

**Comment by Zarel** [comment]
`'lock'` is used for:

- `/lock`, `/unlock` (obviously)
- people locked users can PM (and a lot of similar exceptions, basically only users who can lock are allowed to interact with locked users)
- permission to use scripted buttons
- permission to use `` tags
- not allowed to block PMs
- bypassing `hideBattlesFromTrainerCard`
- seeing help for private commands
- `/sharedbattles`
- `/showglobalpunishments`
- `/globalpunishments`
- seeing hidden roomauth in `/userauth`
- punishment notifications (to prevent spam, if someone in a private room roombans/warns someone outside of the private room, it will only notify if the action is done by a global)
- `/unbattleban`
- `/renamegroupchat`
- bypassing staff impersonation filter for status message
- `<<view-filters>>`
- `/chatlog`
- various helpticket permissions
- viewing IPs in modlog

**Comment by whalemer** [comment]
can('mute') is used for:
- bypassing banwords
- seeing staffintros
- seeing punishments in `/whois`
- /a lot/ of generic staff-only commands (approvals, show, subroomgroupchats, showing secret subrooms, chatlogs, a lot of room activity commands)
- /a lot/ of generic moderation commands (hidetext, showblacklist, banwords list)
- seeing whitelisted users in auth
- helpfilter
- getting helpticket notifications (looks like a mistake, honestly)

**Comment by AnnikaCodes** [comment]
I'm not sure that this is necessary anymore - our current method of determining of `isStaff` hasn't been subject to any complaints lately.

**Comment by Zarel** [comment]
Our permissions system is still very inconsistent.

**Comment by mia-pi-git** [comment]
This bug was fixed some time ago and as such, this PR can be closed.

**Comment by monsanto** [comment]
Mia says its closed, so I'm closing.

**Pull request closed without merging**